### PR TITLE
[Sampling] Add `SGLANG_RENORM_DETERMINISTIC` for bit-identical top-p / top-k renorm

### PR DESCRIPTION
**Pull request by gilfordting**
## Summary

- Add `SGLANG_RENORM_DETERMINISTIC` (default off) routing top-p / top-k renorm to kernels that are bit-identical call to call
- Default stays on the fast flashinfer kernels: the deterministic top-k fallback costs 4-12x, and #33614 already keeps TP ranks in agreement by broadcasting every speculative decision from rank 0 (`SGLANG_SPEC_TP_SYNC=all`)
- Set to `1` when the renorm output itself must agree across TP ranks or be reproducible run to run (logprobs, sampling masks); `SGLANG_SYNC_TOKEN_IDS_ACROSS_TP=1` is the other way to close the gap on the `min_p` path

## Root cause

- `sgl_kernel.top_p_renorm_probs` / `top_k_renorm_probs` forward to flashinfer kernels that pool partial sums with float `atomicAdd`, so byte-identical input returns probabilities differing in the last bits
  - top-p: rows > 2048 dispatch to the AIR radix kernel (flashinfer >= 0.6.7, flashinfer-ai/flashinfer#2752); its default `is_deterministic=False` histogram is a float atomic
  - top-k: `RadixTopKRenormProbMultiCTA` accumulates the kept mass with a float atomic; flashinfer has no deterministic option (fix in flight: flashinfer-ai/flashinfer#5034)
- Every TP rank renormalizes independently; in non-greedy verification (`tree_speculative_sampling_target_only`) the output is compared against a per-position coin and the bonus token is drawn from `relu(q - p)`, so a few-ULP gap between ranks can flip one rank's accept/bonus decision and desynchronize per-rank KV/radix state (#33549, #33289)
- The `min_p` path of `Sampler` has the same exposure and no broadcast (`SGLANG_SYNC_TOKEN_IDS_ACROSS_TP` defaults off)

## Measurements

Kernel-level, 200 repeated calls on the same input (256 rows x 163840 vocab, top_p 0.95 / top_k 40), flashinfer 0.6.18; identical on H100, B200 and B300:

| kernel | calls differing from call 0 | max abs diff |
|---|---|---|
| `top_p_renorm_probs` default (AIR) | 199 / 199 | 1.2e-6, up to ~500K elements, some flip between 0 and non-zero (the cutoff set itself changes) |
| `top_p_renorm_probs(is_deterministic=True)` | 0 / 199 | 0 |
| `top_k_renorm_probs` default (radix multi-CTA) | 199 / 199 | 8.9e-8 (3 ULPs) |
| sgl_kernel's compiled single-CTA `top_k_renorm_probs` | 0 / 199 | 0 |
| torch softmax/topk, `tree_speculative_sampling_target_only`, flashinfer `top_k_top_p_sampling_from_probs`, `sampling_from_probs`, `min_p_sampling_from_probs` (seeded) | 0 | 0 |

The top-p noise is entropy-gated (bit-stable on 4,096 peaky real Llama-3.1-8B decode rows, fires on flat rows); the top-k noise is not (87 to 99 of 99 calls differ even on peaky rows).

End to end, TP=2, DFlash, Llama-3.1-8B + `z-lab/LLaMA3.1-8B-Instruct-DFlash-UltraChat`, `temperature=1.0`, `top_p=0.9`, 32 concurrent multi-turn streams, fp8 KV, with a per-verify-step all-gather of fingerprints of `{rng state, prefix_lens, draft_tokens, target_logits, target_probs, accept_len, bonus, commit_lens, out_tokens}`:

| run | change | outcome |
|---|---|---|
| fast kernels, spec TP sync narrowed, H100 (x2) and B300 (x1) | none | all three hung within 3 to 4 minutes. Divergence chain every time: `target_probs` differs (logits, prefix_lens, RNG identical) -> `bonus`/`out_tokens` differ on one rank -> next step's `draft_tokens` differ -> wedge. 0.4 to 0.7 % of `top_p_renorm_probs` calls returned different bytes when called twice on the same input on the same rank |
| `top_p=1.0` | kernel never called | 15 min, 0 divergence, no hang |
| `top_p=0.9`, `top_k=40` | renorm on `[rows, 40]` uses the single-CTA kernel | 15 min, 26,000 calls, 0 non-deterministic, 0 divergence, no hang |
| `SGLANG_RENORM_DETERMINISTIC=1` | | 30 min, 21,294 requests, 82,000 calls, 0 non-deterministic, 0 divergence, no hang |

Synthetic test traffic had not reproduced this because it used `top_p=1.0`, which skips the renorm kernel entirely.

## Modifications

- `python/sglang/srt/layers/sampling_renorm.py`: `top_p_renorm_prob` / `top_k_renorm_prob` with `deterministic: Optional[bool] = None`, defaulting to `SGLANG_RENORM_DETERMINISTIC`
  - deterministic top-p: flashinfer `top_p_renorm_probs(..., is_deterministic=True)`; without flashinfer (MUSA) the single-CTA kernel compiled into `sgl_kernel`
  - deterministic top-k: the single-CTA `top_k_renorm_probs` compiled into `sgl_kernel` (fixed-order block reductions); goes away once the flashinfer pin includes flashinfer-ai/flashinfer#5034
  - default (`0`) keeps today's kernels
- `sampler.py`, `speculative/dflash_utils.py`, `speculative/eagle_utils.py`: the CUDA and MUSA renorm imports go through that module (the only three call sites)
- `environ.py`: `SGLANG_RENORM_DETERMINISTIC = EnvBool(False)` next to the other determinism knobs
- `sampler.py` (`_sync_token_ids_across_tp`) and `speculative/spec_tp_sync.py`: comments at the two TP-sync sites name the renorm kernels as a known non-deterministic source and point at the knob
- `test/registered/kernels/ops/moe/test_renorm.py`: existing correctness tests run through the new module; two regression tests call each deterministic renorm 30 times on a flat 256 x 128256 distribution and assert bit-identical output, and check via `envs.SGLANG_RENORM_DETERMINISTIC.override(True)` that the knob reaches the same kernel (the same check against the default flashinfer kernels fails 30 times out of 30)

The fix lives in `sglang` rather than in the `sgl_kernel` wrappers on purpose: `sglang-kernel` ships as a pinned prebuilt wheel, so a wrapper change there would neither be testable in CI nor reach users until the next kernel release.

## Speed

Microbenchmark, B200, flashinfer 0.6.18, 50 iterations after warmup, batch x vocab:

| kernel | 32 x 128256 | 256 x 128256 | 256 x 163840 | 989 x 128256 |
|---|---|---|---|---|
| top-p AIR default | 0.136 ms | 0.573 ms | 0.701 ms | 1.87 ms |
| top-p AIR `is_deterministic=True` | 0.165 ms | 0.826 ms | 1.03 ms | 2.71 ms |
| top-k radix default | 0.039 ms | 0.224 ms | 0.268 ms | 0.759 ms |
| top-k single-CTA | 0.482 ms | 0.927 ms | 1.29 ms | 3.04 ms |

Cost of `SGLANG_RENORM_DETERMINISTIC=1` per decode step: renorm runs once per speculative verify step over `batch_size x draft_tokens` rows, and in the plain sampler only on the `min_p` path (the fused `top_k_top_p_sampling_from_probs` kernel is untouched). For 32 requests x 8 draft tokens = 256 rows, 163840 vocab, 25 to 40 ms per step:

| traffic | extra per step | share of step |
|---|---|---|
| `top_p < 1`, no `top_k` (dense path, top-p only) | +0.33 ms | ~1 % |
| batch mixes `top_k` and non-`top_k` requests (dense path, top-k then top-p) | +1.35 ms | ~3 to 5 % |
| every request has `top_k` (sparse `torch.topk` path, renorm on `[rows, k]`) | 0 (already the single-CTA kernel) | 0 |
| greedy or `top_p = 1.0` | 0 (renorm not called) | 0 |

The single-CTA top-k kernel barely scales with row count (0.48 ms at 32 rows vs 0.93 ms at 256), so small batches pay most of it. With the default off there is no change to any path.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :x: [Run #35407882752](https://github.com/sgl-project/sglang/actions/runs/35407882752)
Latest PR Test (Extra): :x: [Run #35407882545](https://github.com/sgl-project/sglang/actions/runs/35407882545)
Latest PR Test (AMD ROCm 10): :x: [Run #35407882639](https://github.com/sgl-project/sglang/actions/runs/35407882639)

### CODE REVIEW

**Comment by gilfordting** [comment]
/rerun-test registered/kernels/ops/moe/test_renorm.py registered/sampling/test_sampling_mask.py registered/core/test_basic_sanity_dflash.py registered/core/test_basic_sanity_dspark.py registered/core/test_basic_sanity_eagle3.py registered/spec/dflash/test_dflash.py registered/spec/eagle/test_eagle_reject_sampling.py registered/spec/eagle/test_spec_eagle.py registered/spec/dspark/test_dspark_kernel_parity.py

**Comment by gilfordting** [comment]
/rerun-test test_metrics.py test_radix_cache_hit.py test_unified_radix_cache_kl_full.py test_hicache_variants.py test_hicache_storage.py

**Comment by gilfordting** [comment]
/rerun-test registered/kernels/ops/moe/test_renorm.py registered/sampling/test_sampling_mask.py registered/core/test_basic_sanity_dflash.py registered/core/test_basic_sanity_dspark.py registered/core/test_basic_sanity_eagle3.py registered/spec/dflash/test_dflash.py registered/spec/eagle/test_eagle_reject_sampling.py registered/spec/eagle/test_spec_eagle.py registered/spec/dspark/test_dspark_kernel_parity.py

**Comment by gilfordting** [comment]
/rerun-test registered/kernels/ops/moe/test_renorm.py registered/sampling/test_sampling_mask.py registered/core/test_basic_sanity_dflash.py registered/core/test_basic_sanity_dspark.py registered/core/test_basic_sanity_eagle3.py registered/spec/dflash/test_dflash.py registered/spec/eagle/test_eagle_reject_sampling.py registered/spec/eagle/test_spec_eagle.py registered/spec/dspark/test_dspark_kernel_parity.py

**Comment by hnyls2002** [comment]
/rerun-test registered/kernels/ops/moe/test_renorm.py registered/sampling/test_sampling_mask.py registered/sampling/test_pytorch_sampling_backend.py registered/spec/dflash/test_dflash.py registered/spec/eagle/test_eagle_reject_sampling.py registered/spec/eagle/test_spec_eagle.py registered/spec/dspark/test_dspark_kernel_parity.py registered/core/test_basic_sanity_dflash.py registered/core/test_basic_sanity_dspark.py
